### ISSUE TITLE
Catalog price rules are not behaving as expected

### ISSUE THREAD

**Issue by BlacklineCloud** [PrestaShop/PrestaShop#35122]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

In my store, I allow local and international orders. For both cases, I offer free shipping and logistics costs are included in the price. Since international logistics prices are higher than local logistics prices, I set the base price of my products globally (where international logistics costs are included) and then add "specific prices" for local buyers on a country-specific basis.

The problem is; I wanted to define a discount for a customer group. I added the catalog price rule. So far everything is normal. This is what I noticed when I checked with my test customer account.

If more than one address is registered in the customer's account (in the current case, there was one local address and one international address), the catalog price rule does not calculate correctly.

For example; If the "specific price" is $10 (for local buyers) and the base price is $20 (for international buyers), it calculates the price at $20 even if the local address is selected in the shopping cart. 

In short, when the product is added to the cart, it should be 10$, but since I have defined a 10 percent discount for purchases of 2 or more items, it should be calculated as 9x2=18$, but it is calculated as 18x2=36$

### Steps to reproduce

Add a catalog price rule with 10% discount for 2 or more items.
Add a specific price rule in a product for specific country.
Add one local and another international address to customer account.
Check discounts on cart.

### PrestaShop version(s) where the bug happened

8.1.2

### PHP version(s) where the bug happened

8.2

**Comment by AureRita** [comment]
Hi @BlacklineCloud ,

Thank you for your report, I reproduce the issue with PrestaShop version 8.1.x, as you can see : 

[recording(68).webm](https://github.com/PrestaShop/PrestaShop/assets/100699445/a9c6f62c-73ef-48a7-9a77-8a3a99c6316e)

I'll add this to the backlog so it can be fixed.

Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank you for your first issue ! This one is clear and we can reproduce the issue on the first try ! 

On the video, we see that two discount doesn't seems to fit. 

Thanks

**Comment by boo-code** [comment]
The arithmetic in the report is reproducible from the resolution query alone, without a browser, and it is not about which address is selected.

`SpecificPrice::getSpecificPrice()` returns **one** row, ordered by

```sql
ORDER BY `id_product_attribute` DESC, `id_cart` DESC, `from_quantity` DESC,
         `id_specific_price_rule` ASC, `score` DESC, `to` DESC, `from` DESC
```

`from_quantity` is ranked **above** `score`, and `score` is the only place where `PS_SPECIFIC_PRICE_PRIORITIES` has any effect. So a quantity tier outranks the merchant-configured priority order, whatever that order is.

Measured on `9.2.x` with the shipped default priorities (`id_group;id_currency;id_country;id_shop`), a product priced 20 with two competing rows — a country price of 10 at `from_quantity = 1`, and a catalog price rule of 10% at `from_quantity = 2` with `price = -1` — resolved for a cart in that country:

```
quantity = 1
 id_specific_price | id_specific_price_rule | id_country | price | from_quantity | reduction | score
               354 |                      0 |         21 | 10.00 |             1 |      0.00 |    30

quantity = 2
 id_specific_price | id_specific_price_rule | id_country | price | from_quantity | reduction | score
               355 |                    999 |          0 | -1.00 |             2 |      0.10 |    26   <- returned
               354 |                      0 |         21 | 10.00 |             1 |      0.00 |    30
```

At quantity 2 the catalog rule row wins with the **lower** score, because `from_quantity DESC` is compared first. Its `price` is `-1`, which means "keep the base price", and `Product::priceCalculation()` then takes that base from the product itself:

```php
if (!$specific_price || $specific_price['price'] < 0) {
    $price = (float) $result['price'];      // 20, not the country's 10
}
```

so the country price is unreachable and the result is `(20 - 10%) x 2 = 36`, exactly the number in the report. The `9x2 = 18` the reporter expects would require the country row's base and the rule's reduction to be combined, and `getSpecificPrice()` returns a single row, so no code path can express that today.

Worth noting what the obvious fix would do: moving `from_quantity` after `score` makes the country row win at quantity 2, giving `10 x 2 = 20`. That is not the expected result either, and it would stop volume discounts from applying to any product that has a country-, group- or currency-scoped price — a wider regression than the bug. The behaviour that matches the report is composition (highest-scoring row for the base price, quantity row for the reduction), which is a change to the single-row contract of `getSpecificPrice()` rather than a reordering, so it probably belongs to #20183.

Two smaller observations while measuring: the priority list configured in the back office is silently overridden by any quantity tier, which is worth stating in the UI or the docs whatever is decided; and the multiple-addresses condition in the report is not required — `Product::getPriceStatic()` takes the country from `$cart->{PS_TAX_ADDRESS_TYPE}`, so the selected address is used correctly, and the same result appears with a single address as soon as the quantity reaches the rule's threshold.
